registry/storage/driver/middleware/storage_fallback/middleware.go

```diff
@@ -0,0 +1,67 @@
+package storage_fallback
+
+import (
+	"context"
+	"fmt"
+	dcontext "github.com/docker/distribution/context"
+	storagedriver "github.com/docker/distribution/registry/storage/driver"
+	"github.com/docker/distribution/registry/storage/driver/factory"
+	storagemiddleware "github.com/docker/distribution/registry/storage/driver/middleware"
+	"io"
+)
+
+type fallbackStorageDriver struct {
+	storagedriver.StorageDriver
+	Fallback storagedriver.StorageDriver
+}
+
+func (sd *fallbackStorageDriver) Name() string {
+	return fmt.Sprintf("%v -> %v", sd.StorageDriver.Name(), sd.Fallback.Name())
+}
+
+func (sd *fallbackStorageDriver) GetContent(ctx context.Context, path string) ([]byte, error) {
+	result, err := sd.StorageDriver.GetContent(ctx, path)
+	if err != nil {
+		switch err.(type) {
+		case storagedriver.PathNotFoundError:
```

Comment:
why not use fallback when you hit PathNotFoundError?

Reply:
my initial thinking was that I just wanted the middleware to handle 5xx-like errors, but I guess it can't hurt to have to fallback enabled for everything

Reply:
For PathNotFoundError, i think we should def be graceful about it. Imagine if S3 writes are down and we need to publish a new docker image to workaround it during a critisit. Ideally, the build would still get published to GCS and things would gracefully deal with getting it from GCS instead of s3 if not found in s3.